cpp/src/arrow/compute/kernels/vector_sort_test.cc

```diff
@@ -1936,6 +1940,95 @@ TEST_P(TestTableSortIndicesRandom, Sort) {
   }
 }
 
+// Ranking Tests
+//
+TEST(ArrayRankFunction, Array) {
+  auto arr = ArrayFromJSON(int16(), "[0, 1, -3, -42, 5]");
+  auto expectedAsc = ArrayFromJSON(uint64(), "[3, 4, 2, 1, 5]");
+  for (auto null_placement : AllNullPlacements()) {
+    for (auto tiebreaker : AllTiebreakers()) {
+      RankOptions options(SortOrder::Ascending, null_placement, tiebreaker);
+      ASSERT_OK_AND_ASSIGN(auto actual, CallFunction("rank", {arr}, &options));
+      AssertDatumsEqual(expectedAsc, actual, /*verbose=*/true);
+    }
+  }
+
+  auto expectedDesc = ArrayFromJSON(uint64(), "[3, 2, 4, 5, 1]");
+  for (auto null_placement : AllNullPlacements()) {
+    for (auto tiebreaker : AllTiebreakers()) {
+      RankOptions options(SortOrder::Descending, null_placement, tiebreaker);
+      ASSERT_OK_AND_ASSIGN(auto actual, CallFunction("rank", {arr}, &options));
+      AssertDatumsEqual(expectedDesc, actual, /*verbose=*/true);
+    }
+  }
+}
+
+TEST(ArrayRankFunction, NullHandling) {
+  auto arr = ArrayFromJSON(int16(), "[0, 1, null, null, 2]");
+  auto expectedEnd = ArrayFromJSON(uint64(), "[1, 2, 4, 5, 3]");
+  RankOptions optionsEnd(SortOrder::Ascending, NullPlacement::AtEnd, Tiebreaker::First);
+  ASSERT_OK_AND_ASSIGN(auto actualEnd, CallFunction("rank", {arr}, &optionsEnd));
+  AssertDatumsEqual(expectedEnd, actualEnd, /*verbose=*/true);
+
+  auto expectedStart = ArrayFromJSON(uint64(), "[3, 4, 1, 2, 5]");
+  RankOptions optionsStart(SortOrder::Ascending, NullPlacement::AtStart,
+                           Tiebreaker::First);
+  ASSERT_OK_AND_ASSIGN(auto actualStart, CallFunction("rank", {arr}, &optionsStart));
+  AssertDatumsEqual(expectedStart, actualStart, /*verbose=*/true);
+}
+
+TEST(ArrayRankFunction, TiebreakHandlingAsc) {
```

Comment:
Can you add nuls into the mix? I wonder if multiple nulls should be considered ties (AFAIU, `NULL == NULL` is false in SQL-land). @lidavidm What do you think?

Reply:
I guess Pandas has options for this:

```
na_option{‘keep’, ‘top’, ‘bottom’}, default ‘keep’

    How to rank NaN values:

        keep: assign NaN rank to NaN values

        top: assign lowest rank to NaN values

        bottom: assign highest rank to NaN values
```

Reply:
Hmm, according to some examples on the Web, even SQL considers NULLs identical for the purposes of RANK().

Reply:
Hmm, perhaps we also don't need to implement all possible options at once? :-)

Reply:
That's what I am thinking as well - might be worth a separate PR / initiative for the Keep scenario